### ISSUE TITLE
Document Kotlin Support

### ISSUE THREAD

**Issue by NfNitLoop** [smallrye/smallrye-graphql#1853]
Hello! Thanks for this project! We're enjoying using it.

We'd like to use Kotlin to define our GraphQL models and implementation. But I don't see documentation about what Kotlin features are supported or not supported by SmallRye.

I think I've figured the list below out by trial an error, but it would be nice to have them documented somewhere as explicitly supported/unsupported. Can someone confirm these? 😊 

* SmallRye uses JSON-B for deserialization, so if you use any data classes as inputs, they'll need a `@JsonbCretor` annotation annotation on their constructor.  (example below)
* SmallRye supports Kotlin's nullability annotations and will properly propagate them through schema generation.  ([code])
* Unlike what I read [elsewhere], I don't seem to need to annotate each constructor parameter with `@JsonbProperty("names")`, that seems to be handled automatically.   (Is this handled by SmallRye? Or its JSON-B library? Is that a feature I can rely on?)
* SmallRye doesn't support Kotlin default parameters when deserializing objects. 

[code]: https://github.com/smallrye/smallrye-graphql/blob/f7c14bcf24d7ac53950eff7119a216ece2ee3fa6/common/schema-builder/src/main/java/io/smallrye/graphql/schema/helper/NonNullHelper.java#L75
[elsewhere]: https://apiumhub.com/tech-blog-barcelona/json-in-kotlin-comparing-options/#JSON-B

```kotlin
// Sample kotlin code showing required annotations.
data class Foo @JsonbCreator constructor(
    val name: String,
    val bar: Bar?,
)

data class Bar @JsonbCreator constructor(
    val name: String
)

@GraphQLApi
class MyService {
    @Query("foo")
    @Description("GQL Docs here")
    fun getFoo() = Foo(
            name = "My foo",
            bar = Bar(
                name = "Hello from Bar!"
            ),
    )

    @Mutation("foo")
    // Nice: I get to use Foo as an Input type too. ❤️
    fun setFoo(foo: Foo): String {
        val msg = "Got foo: $foo"
        println(msg)
        return msg
    }
}
```

**Comment by t1** [comment]
@NfNitLoop: thanks for your input. Would you like to turn this into a PR for a new page in the `docs` folder?

You don't have to add `@JsonProperty` to parameters, probably because we use the parameter names that can be compiled into the bytecode; `javac` does this with the `-parameters` flag, maybe Kotlin does this out of the box.

**Comment by NfNitLoop** [comment]
I'm not quite ready to turn it into a PR w/o knowing answers to some of the questions I asked:

* Is the automatic handling of `data class` constructor field names a feature SmallRye intends to support?

I guess the main underlying question is, in general: What level of Kotlin compatibility does the project *aim* to provide? I can't answer that as someone who is not a main contributor. But it would be nice to have that stated somewhere so folks know what to expect. 😊

**Comment by phillip-kruger** [comment]
Hi @NfNitLoop - none of the main contributors know Kotlin. SmallRye GraphQL has no intention w.r.t Kotlin, we do what we can with supporting Kotlin. If currently data class constructor field names is NOT supported, then you are most welcome to open a PR that adds support for it. We will definitely consider it. As long as a PR does not add a bunch of new dependencies or rewrite a bunch of things, there should be no reason not to accept a feature. This project can do with a contributor that has knowledge and passion for Kotlin...

**Comment by jmartisk** [comment]
Yeah. I like Kotlin but I am not very skilled with it. Any contributions to make us support it better are welcome. I can also imagine a Kotlin-based test suite, that would be nice.

**Comment by danielspeixoto** [comment]
Hi @NfNitLoop, having the same issue around here. I am still deciding if we are going to use data classes or not. What you ended up doing?
Thanks.
